channel: Add option to notify users newly added to a channel.

kuv2707 opened the pull request:
When someone adds another user to a channel, we send the user that was added a Notification Bot message to let them know.

This PR adds an option for whether or not this message is sent. If the number of newly added users is more than 100, 
we don't send notifications.

- The checkbox unchecks and becomes disabled if the number of users to be added to the channel crosses the limit (100).
- If the number of users then reduces, the checkbox becomes enabled again, but is left unchecked.


Fixes #31189

[#api design > options for notifying new channel subscribers @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/options.20for.20notifying.20new.20channel.20subscribers/near/2199784)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
<details>
<summary>Screenshots and screen captures</summary>

*I have set the max limit to 2 users for testing*

<img width="1470" alt="image" src="https://github.com/user-attachments/assets/f0044934-13ce-4e7a-aff2-a58ff2201f16" />

<img width="1470" alt="image" src="https://github.com/user-attachments/assets/fc713773-6bf7-4031-91a2-fb77c7fcc674" />


**Narrow width**
<img width="842" alt="image" src="https://github.com/user-attachments/assets/49b37fa2-f8c1-400b-86d3-232df69a1369" />

The cursor changes to <not-allowed> when hovering over the checkbox while it is disabled. 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 commented:
@zulipbot add "buddy review"

@pratik-pc Please have a look

pratik-pc commented:
[Screencast from 05-08-24 05:55:58 PM IST.webm](https://github.com/user-attachments/assets/76985531-a5ab-4d6b-9926-d4c95e05c04c)

@kuv2707 When I add more than 2 members, I can't enable the notification checkbox

kuv2707 commented:
Actually I had set the limit to 2 for testing. Once it passes buddy review, I'll bump it up to 100

pratik-pc commented:
@kuv2707 Oh okay then it is working as intended. Though while creating new stream checkbox is not getting disabled even though works correctly when adding subscribers to existing streams.

kuv2707 commented:
@pratik-pc Things work fine on my side. 

Can you elaborate on what exactly doesn't work correctly when creating a new stream?

pratik-pc commented:
[Screencast from 09-08-24 02:56:45 AM IST.webm](https://github.com/user-attachments/assets/bf8ba71d-235d-483b-8ed7-02985a60b01e)

@kuv2707 Alright I guess this only happens when you create stream when modal is already open.

pratik-pc commented:
[Screencast from 13-08-24 10:30:05 PM IST.webm](https://github.com/user-attachments/assets/53efd302-402c-44e4-b485-28074f5177fd)

Looks go to me now. Just a simple issue with the newly introduced button; inconsistent highlight and clickable area.

kuv2707 commented:
@pratik-pc Please have a look now.

The inconsistency was due to the elements from stream edit ui lingering even after the view changes to stream creation.
The solution was to remove the checkbox from there so that only one instance of it is there in the DOM.

alya commented:
@pratik-pc does this look good to you now?

pratik-pc commented:
Looks good to me now
@zulipbot remove "buddy review"

kuv2707 commented:
@zulipbot add "mentor review"

@Riken-Shah @madrix01 Please have a look

kuv2707 commented:
@alya It would be great if a maintainer took a look at this, since it seems that my mentors wouldn't be able to dedicate time anymore

alya commented:
Yes, thanks for flagging! @sahil839 perhaps you could review this one?

sahil839 commented on web/src/stream_edit_subscribers.ts:
```diff
@@ -263,7 +270,13 @@ function subscribe_new_users({pill_user_ids}: {pill_user_ids: number[]}): void {
+    subscriber_api.add_user_ids_to_stream(
+        user_ids,
+        sub,
+        Boolean($("#send_notification_to_new_subscribers").prop("checked")),
```
Using `is(":checked")` would not require `Boolean`.

kuv2707 replied:
This caused a typescript error otherwise

sahil839 replied:
I don't think it will cause error when using `is(":checked")`.

kuv2707 replied:
Yes, changed.

sahil839 reviewed (commented):
Posted some comments.

saubhagya-patel commented:
Hi, @kuv2707, are you working on this one? If not, could I please carry it forward?
Thank you.

kuv2707 commented:
@saubhagya-patel I am on it.

kuv2707 commented on web/styles/subscriptions.css:
```diff
@@ -709,6 +709,11 @@ h4.user_group_setting_subsection_title {
         flex-wrap: nowrap;
     }
 
+    #send_notification_to_new_subscribers_container.disabled * {
```
Sure, removed.

kuv2707 commented on web/src/stream_settings_components.js:
```diff
@@ -52,6 +52,7 @@ export const show_subs_pane = {
             );
         } else {
             $("#subscription_overlay .stream-info-title").html(render_selected_stream_title({sub}));
+            $("#stream_settings #send_notification_to_new_subscribers_container").remove();
```
I managed to drop it.

kuv2707 commented on web/templates/stream_settings/add_subscribers_form.hbs:
```diff
@@ -1,3 +1,10 @@
+<div class="subsection-parent inline-block" id="send_notification_to_new_subscribers_container">
```
Made it into a class.

kuv2707 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10186,6 +10186,13 @@ paths:
+                    a Notification Bot DM to let them know that they have been added to the channel.
+                  type: boolean
+                  default: true
+                  example: true
```
Added an entry in changelog and bumped version in `version.py`

sahil839 commented on zerver/views/streams.py:
```diff
@@ -738,14 +740,15 @@ def add_subscriptions_backend(
-        created_streams=created_streams,
-        announce=announce,
-    )
+    if send_notification_to_new_users and len(result["subscribed"]) <= MAX_SUBS_FOR_NOTIFICATION:
```
I am not sure but should we raise an error if user passes `send_notification_to_new_users` as `True` when adding more than 100 users.

kuv2707 replied:
Another option is to notify in the Response that notifications weren't sent.

kuv2707 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10355,6 +10355,13 @@ paths:
+                send_notification_to_new_users:
+                  description: |
+                    This parameter determines whether the users newly added to a channel are sent
+                    a Notification Bot DM to let them know that they have been added to the channel.
```
Added.

kuv2707 commented on zerver/views/streams.py:
```diff
@@ -685,14 +687,15 @@ def add_subscriptions_backend(
-        created_streams=created_streams,
-        announce=announce,
-    )
+    if send_notification_to_new_users and len(result["subscribed"]) <= MAX_SUBS_FOR_NOTIFICATION:
```
Added some tests.

kuv2707 commented:
@sahil839 Please have a look.

Summary of changes:
- Added tests
- Added changelog etc.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10574,6 +10585,13 @@ paths:
+                          Whether Notification Bot DMs were sent
+                          to the users newly added to
+                          the channel.
+                          Only present if `send_notification_to_new_users` is `true`.
```
This should also include a "Changes" entry.

kuv2707 replied:
Added.

sahil839 commented on web/src/add_subscribers_pill.ts:
```diff
@@ -103,17 +103,29 @@ export function generate_pill_html(item: CombinedPill): string {
+        pill_widget.onPillRemove(() => {
+            onPillRemoveAction(get_pill_user_ids(pill_widget));
+            // If all the pills are removed, disable the add button.
+            $pill_widget_button.prop("disabled", pill_widget.items().length === 0);
```
This will not disable the buttons for cases where `onPillRemoveAction` will not be passed.

kuv2707 replied:
Oh, the `if` check was supposed to be inside the callback.
Fixed.

sahil839 commented on version.py:
```diff
@@ -34,7 +34,8 @@
 
-API_FEATURE_LEVEL = 337  # Last bumped for voice_only param addition for BigBlueButton
+API_FEATURE_LEVEL = 338  # Last bumped for adding setting
+# `send_notification_to_new_users`.
```
This comment can be added in a single line.

kuv2707 replied:
Done

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10510,6 +10510,17 @@ paths:
+                send_notification_to_new_users:
+                  description: |
+                    This parameter determines whether the users newly added to a channel are sent
+                    a Notification Bot DM to let them know that they have been added to the channel.
```
Would be good to mention that the notification is not sent if number of users added is larger than 100.

kuv2707 replied:
Added a clause for that.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10588,6 +10599,15 @@ paths:
+                        description: |
+                          Whether Notification Bot DMs were sent
+                          to the users newly added to
+                          the channel.
```
The line breaks seem random here.

sahil839 commented on zerver/tests/test_subs.py:
```diff
@@ -756,6 +757,51 @@ def test_acting_user_is_creator(self) -> None:
+
+        data = {
+            "subscriptions": '[{"name":"brand new stream 1","description":""}]',
+            "history_public_to_subscribers": "true",
```
Why we do not use `orjson.dumps(True).decode()` here for boolean fields? I see same pattern used at other places in the same file too, so not sure about that.

kuv2707 replied:
I think that would be too verbose.
In any case, if we want uniformity, we can replace these in a separate PR.

sahil839 reviewed (commented):
Posted a couple of comments for documentation changes which I missed before.

Also, I tested and the notification DM is not being sent even when checkbox is checked. Can you verify?

kuv2707 commented:
> Also, I tested and the notification DM is not being sent even when checkbox is checked. Can you verify?

This only happens when creating channels. If we add users to an existing channel, they do receive a DM.

The reason is mentioned in `zerver/views/streams.py` in function `send_messages_for_new_subscribers`:

```
            # For each user, we notify them about newly subscribed streams, except for
            # streams that were newly created.
            notify_stream_names = set(subscribed_stream_names) - newly_created_stream_names
```

Do we wish to change this behaviour?

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -10725,6 +10744,13 @@ paths:
+                          Whether Notification Bot DMs were sent
+                          to the users newly added to
+                          the channel.
+                          Only present if `send_notification_to_new_users` is `true`.
```
This should also be formatted like above and also needs "Changes" entry.

kuv2707 replied:
Done

sahil839 reviewed (commented):
If we do not send notifications on creating a new channel, then we should not show checkbox in the creation form and remove all the code related to it.

kuv2707 commented:
@sahil839 Please have a look. I have removed the code related to showing the checkbox in stream creation form. Also addressed the last review comment.

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -184,7 +186,9 @@ export function create<ItemType extends {type: string}>(
 
             if (!quiet && store.onPillCreate !== undefined) {
-                store.onPillCreate();
+                for (const f of store.onPillCreate) {
```
I don't get why is this needed? Can't a single function be passed to `onPillCreate` which calls all the needed functions?

kuv2707 replied:
In retrospect, this seems an overkill.

Alternatively, we could extract the currently registered callback and register a new callback, calling the previous one inside it. 
@sahil839 Should I do this instead?

sahil839 replied:
I think would be clean enough to just pass an optional argument `pill_update_callback` to `set_up_handlers_for_add_button_state` and `add_subscribers_pill.create`. And `stream_edit_update_notification_choice` can just compute the user IDs itself instead of receivng it as a parameter as it can already access `pill_widget`.

kuv2707 replied:
Made this change.

sahil839 replied:
These changes can be removed now, right?

sahil839 commented on web/src/state_data.ts:
```diff
@@ -566,7 +566,12 @@ export const state_data_schema = z
+    .and(realm_schema.transform((realm) => ({realm})))
+    .and(
+        z
+            .object({max_subs_for_notification: z.number()})
```
Why is this not done like other similar fields like `max_topic_length`?

sahil839 replied:
Was this discussed whether we should pass this value in `/register` response, or can we just define it in client separately?

kuv2707 replied:
@sahil839  Should I change it back to having a copy separately?

sahil839 replied:
Ok, we can keep it for now and can wait for feedback from Tim before changing. But I think this should be handled just like other fields in this schema so that it can just be accessed from `realm`.

sahil839 commented on web/src/ui_init.js:
```diff
@@ -532,7 +532,7 @@ export function initialize_everything(state_data) {
     stream_edit.initialize();
     user_group_edit.initialize();
-    stream_edit_subscribers.initialize();
+    stream_edit_subscribers.initialize(state_data.max_subs_for_notification);
```
I think we can directly use `realm.max_subs_for_notification` like we do for fields like `max_topic_length`.

sahil839 commented on zerver/views/streams.py:
```diff
@@ -822,10 +822,10 @@ def send_messages_for_new_subscribers(
         for id, subscribed_stream_names in new_subscriptions.items():
             if id == str(user_profile.id):
-                # Don't send a Zulip if you invited yourself.
+                # Don't send a Zulip invitation if you invited yourself.
```
This still seems incorrect. I think thos can be "Don't send a notification DM if you subscribed yourself.".

kuv2707 replied:
Yeah, that makes more sense

sahil839 commented on zerver/views/streams.py:
```diff
@@ -822,10 +822,10 @@ def send_messages_for_new_subscribers(
                 continue
             if bots[id]:
-                # Don't send invitation Zulips to bots
+                # Don't send Zulip invitations to bots
```
Similarly this can be "Don't send notification DMs to bots"

kuv2707 replied:
Changed

sahil839 reviewed (commented):
Posted few comments.

kuv2707 commented:
@sahil839 I have addressed the feedback. Please have a look.

sahil839 commented on web/src/stream_edit_subscribers.ts:
```diff
@@ -30,6 +30,8 @@ import type {CombinedPillContainer} from "./typeahead_helper.ts";
 import * as user_groups from "./user_groups.ts";
 import * as user_sort from "./user_sort.ts";
 
+export let max_subs_for_notification = 2;
```
Was this done for some testing?

kuv2707 replied:
It was also meant to be an init value for the setting, though no longer required.

sahil839 commented on web/src/stream_edit_subscribers.ts:
```diff
@@ -99,6 +101,23 @@ function show_stream_subscription_request_result({
+export function update_notification_choice_checkbox(added_user_count: number): void {
+    const $send_notification_checkbox = $(".send_notification_to_new_subscribers");
+    const $send_notification_container = $(".send_notification_to_new_subscribers_container");
+    if (added_user_count > max_subs_for_notification) {
```
This can be changed to `realm.max_subs_for_notification`.

kuv2707 replied:
Done

sahil839 commented on web/src/stream_edit_subscribers.ts:
```diff
@@ -478,4 +504,5 @@ export function initialize(): void {
             remove_subscriber({stream_id, target_user_id, $list_entry});
         },
     );
+    max_subs_for_notification = max_subs;
```
This should not be needed I guess. You can directly use `realm.max_subs_for_notification`.

kuv2707 replied:
done

kuv2707 commented:
@sahil839 Addressed the feedback. I've shifted to using `realm.max_sub_for_notification`, and remove the changes to `input_pill.ts`

sahil839 commented:
Looks good to me. @alya do you want to have a look at this once?

alya commented:
Thanks! Looks good overall, but let's tweak a few things:

1. Let's just put the setting on the line below the heading. I think that will look best in general, not just on narrow screens.
2. Message text should be, e.g.:

> @_**Desdemona** subscribed you to [#devel](http://localhost:9991/#narrow/channel/11-devel).

So, silent mention, and I don't think it needs to say "the channel".

3. Setting label tweak:
> Send notification message to newly subscribed users

3. Could you please document this on https://zulip.com/help/subscribe-users-to-a-channel ? Like this:

> 6. Configure **Send notification message to newly subscribed users** as desired.

alya commented:
Looks good to me in manual testing otherwise! You can tag for integration review once the comment above has been addressed.

alya commented:
What happens if a limited guest user gets added to a channel by someone whose name they don't have access to? (I haven't checked if this was already discussed above.)

kuv2707 commented:
@alya Thanks for the feedback.
- Changed the label text.
- Configured the setting to always be below the heading.
- Added the help center bullets in a commit.

Regarding making changes to the message sent by the bot, it would be best to address it in a followup PR, since it is fairly independent of this issue.

kuv2707 commented:
@zulipbot add "integration review"

kuv2707 commented:
@zulipbot remove "maintainer review"

alya commented:
Thanks! The screenshots in the PR description seem to be out of date now?

kuv2707 commented:
Oh, updated!

alya commented:
I guess the narrow one is still stale?

alya commented:
Maybe we should change the tooltip to be "... when subscribing more than N users."

kuv2707 commented:
>I guess the narrow one is still stale?

I thought it was OK to leave it so, since nothing in the setting label changed there (though I now see that the surrounding context did change a bit). I've updated the screenshots.

>Maybe we should change the tooltip to be "... when subscribing more than N users."
Changed!


@alya I hope this is fine now.

kuv2707 commented:
When disabling the option, we simply add the number of users in user groups, ignoring the fact that some may already be subscribed to the channel. The reasoning is that the user *did* want to add more people than allowed, and we prohibited that.

Should we change this and compute the exact number of new subscribers?

kuv2707 commented:
> What happens if a limited guest user gets added to a channel by someone whose name they don't have access to? (I haven't checked if this was already discussed above.)
> 
> I guess if the message just says "Unknown user" , that's fine. Not worth doing something special for that edge case, so long as it isn't broken.

The message does show their name, but the user card says "Unknown user". 
<img width="516" alt="image" src="https://github.com/user-attachments/assets/0aa7c60e-d78d-443f-94c0-75315aeb504a" />


This is consistent with when the inaccessible user is mentioned in a channel message or DM.

<img width="576" alt="image" src="https://github.com/user-attachments/assets/a342311d-e364-46b3-9886-926893efad31" />

alya commented:
> Should we change this and compute the exact number of new subscribers?

I'll let @timabbott comment, but the current approach seems fine to me.

alya commented:
@timabbott The screenshots look good to me now.

kuv2707 commented:
I had to make few minor changes to accomodate #34503.

kuv2707 commented:
@sahil839 Please have a look.

timabbott commented:
This lgtm except I'm not sure the API changes are approved via the API design channel. (The names might need discussion and improvement if not).

timabbott commented on api_docs/unmerged.d/ZF-06b4d1.md:
```diff
@@ -0,0 +1,13 @@
+
+* [`POST /users/me/subscriptions`](/api/get-subscriptions): Added
+`new_subscription_messages_sent` to the response, which is only present if
+the parameter `send_new_subscription_messages` in the request is `true`.
```
I'll do some wording tweaks too, but these links are broken. I think the issue here is probably that we don't offer a way to view these to make it easy to test. @Vector73 probably we should try to make a /api/changelog/unmerged page that includes lets you preview these entries so this mistake becomes not easy to make.

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 54850df953..d6778e5fd7 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -22,14 +22,12 @@ format used by the Zulip server that they are interacting with.
 
 **Feature level 397**
 
-**Feature level 385**
-
-* [`POST /users/me/subscriptions`](/api/get-subscriptions): Added parameter
+* [`POST /users/me/subscriptions`](/api/subscribe): Added parameter
 `send_new_subscription_messages` which determines whether the users newly added
 to a channel are sent a Notification Bot DM to let them know that they have been
 added to the channel.
 
-* [`POST /users/me/subscriptions`](/api/get-subscriptions): Added
+* [`POST /users/me/subscriptions`](/api/subscribe): Added
 `new_subscription_messages_sent` to the response, which is only present if
 the parameter `send_new_subscription_messages` in the request is `true`.
 
```

Vector73 replied:
I think showing the updates on "/api/changelog" itself would be the better option (adding a new page with just some updates seems redundant) -- opened #35016.

timabbott commented:
I made these changes to the backend commit:

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 54850df953..f8322cc96d 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -22,16 +22,14 @@ format used by the Zulip server that they are interacting with.
 
 **Feature level 397**
 
-**Feature level 385**
-
-* [`POST /users/me/subscriptions`](/api/get-subscriptions): Added parameter
-`send_new_subscription_messages` which determines whether the users newly added
-to a channel are sent a Notification Bot DM to let them know that they have been
-added to the channel.
+* [`POST /users/me/subscriptions`](/api/subscribe): Added parameter
+  `send_new_subscription_messages` which determines whether the user
+  would like Notification Bot to notify other users who the request
+  adds to a channel.
 
-* [`POST /users/me/subscriptions`](/api/get-subscriptions): Added
-`new_subscription_messages_sent` to the response, which is only present if
-the parameter `send_new_subscription_messages` in the request is `true`.
+* [`POST /users/me/subscriptions`](/api/subscribe): Added
+  `new_subscription_messages_sent` to the response, which is only
+  present if `send_new_subscription_messages` was `true` in the request.
 
 * [`POST /register`](/api/register-queue): Added `max_bulk_new_subscription_messages`
   to the response.
diff --git a/zerver/lib/events.py b/zerver/lib/events.py
index aeaf9d016a..f30da779aa 100644
--- a/zerver/lib/events.py
+++ b/zerver/lib/events.py
@@ -551,7 +551,7 @@ def fetch_initial_state_data(
 
         state["max_stream_name_length"] = Stream.MAX_NAME_LENGTH
         state["max_stream_description_length"] = Stream.MAX_DESCRIPTION_LENGTH
-        state["max_bulk_new_subscription_messages"] = streams.MAX_BULK_NEW_SUBSCRIPTION_MESSAGES
+        state["max_bulk_new_subscription_messages"] = settings.MAX_BULK_NEW_SUBSCRIPTION_MESSAGES
         state["max_topic_length"] = MAX_TOPIC_NAME_LENGTH
         state["max_message_length"] = settings.MAX_MESSAGE_LENGTH
         if realm.demo_organization_scheduled_deletion_date is not None:
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 29550201d1..36e73cc7d8 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -11598,15 +11598,18 @@ paths:
                   example: 1
                 send_new_subscription_messages:
                   description: |
-                    This parameter determines whether the users newly added to a
-                    channel are sent a Notification Bot DM to let them know that
-                    they have been added to the channel. A DM is not sent if the
-                    number of users added is more than 100 irrespective of this
-                    parameter.
-
-                    **Changes**: Before Zulip 11.0 (feature level 397), all newly
-                    added users were automatically sent a Notification Bot DM,
-                    which was costly when adding a large number of users.
+                    Whether any other users newly subscribed via this request should be
+                    sent a Notification Bot DM notifying them about their new
+                    subscription.
+
+                    The server will never send Notification Bot DMs if more than
+                    `max_bulk_new_subscription_messages` (available in the [`POST
+                    /register`](/api/register-queue) response) users were subscribed in
+                    this request.
+
+                    **Changes**: Before Zulip 11.0 (feature level 397), new subscribers
+                    were always sent a Notification Bot DM, which was unduly expensive
+                    when bulk-subscribing thousands of users to a channel.
                   type: boolean
                   default: true
                   example: true
@@ -11703,10 +11706,13 @@ paths:
                       new_subscription_messages_sent:
                         type: boolean
                         description: |
-                          Whether Notification Bot DMs were sent to the users
-                          newly added to the channel. It is only present if
-                          the parameter `send_new_subscription_messages` in the
-                          request is `true`.
+                          Only present if the parameter `send_new_subscription_messages`
+                          in the request was `true`.
+
+                          Whether Notification Bot DMs in fact sent to the added
+                          subscribers as requested by the `send_new_subscription_messages`
+                          parameter. Clients may find this value useful to communicate
+                          with users about the effect of this request.
 
                           **Changes**: New in Zulip 11.0 (feature level 397).
                     example:
@@ -11848,10 +11854,13 @@ paths:
                       new_subscription_messages_sent:
                         type: boolean
                         description: |
-                          Whether Notification Bot DMs were sent
-                          to the users newly added to the channel.
-                          Only present if `send_new_subscription_messages` is
-                          `true`.
+                          Only present if the parameter `send_new_subscription_messages`
+                          in the request was `true`.
+
+                          Whether Notification Bot DMs in fact sent to the added
+                          subscribers as requested by the `send_new_subscription_messages`
+                          parameter. Clients may find this value useful to communicate
+                          with users about the effect of this request.
 
                           **Changes**: New in Zulip 11.0 (feature level 397).
                     example:
@@ -20117,9 +20126,9 @@ paths:
                               $ref: "#/components/schemas/GroupPermissionSetting"
                       max_bulk_new_subscription_messages:
                         description: |
-                          Maximum number of subscribers that can be
-                          added to a channel at once such that they
-                          receive a Notification Bot DM.
+                          Maximum number of new subscribers for which the server will
+                          respect the `send_new_subscription_messages` parameter when
+                          [adding subscribers to a channel](/api/subscribe#parameter-send_new_subscription_messages).
 
                           **Changes**: New in Zulip 11.0 (feature level 397).
                         type: number
diff --git a/zerver/tests/test_subs.py b/zerver/tests/test_subs.py
index dd87f54aa0..63a46144b4 100644
--- a/zerver/tests/test_subs.py
+++ b/zerver/tests/test_subs.py
@@ -87,7 +87,7 @@ from zerver.models.realm_audit_logs import AuditLogEventType
 from zerver.models.realms import get_realm
 from zerver.models.streams import StreamTopicsPolicyEnum, get_default_stream_groups, get_stream
 from zerver.models.users import active_non_guest_user_ids, get_user, get_user_profile_by_id_in_realm
-from zerver.views.streams import compose_views, set_max_subs_for_notification
+from zerver.views.streams import compose_views
 
 if TYPE_CHECKING:
     from django.test.client import _MonkeyPatchedWSGIResponse as TestHttpResponse
@@ -4945,11 +4945,10 @@ class SubscriptionAPITest(ZulipTestCase):
         data = self.assert_json_success(response)
         self.assertEqual(data["new_subscription_messages_sent"], True)
 
-        set_max_subs_for_notification(5)
-
-        response = self.subscribe_via_post(
-            desdemona, ["Test stream 2"], dict(principals=orjson.dumps(user_ids).decode())
-        )
+        with self.settings(MAX_BULK_NEW_SUBSCRIPTION_MESSAGES=5):
+            response = self.subscribe_via_post(
+                desdemona, ["Test stream 2"], dict(principals=orjson.dumps(user_ids).decode())
+            )
         data = self.assert_json_success(response)
         self.assertEqual(data["new_subscription_messages_sent"], False)
 
diff --git a/zerver/views/streams.py b/zerver/views/streams.py
index 6974cb2fc4..97186dfbd2 100644
--- a/zerver/views/streams.py
+++ b/zerver/views/streams.py
@@ -632,12 +632,6 @@ def you_were_just_subscribed_message(
 
 
 RETENTION_DEFAULT: str | int = "realm_default"
-MAX_BULK_NEW_SUBSCRIPTION_MESSAGES: int = 100
-
-
-def set_max_subs_for_notification(value: int) -> None:
-    global MAX_BULK_NEW_SUBSCRIPTION_MESSAGES
-    MAX_BULK_NEW_SUBSCRIPTION_MESSAGES = value
 
 
 @transaction.atomic(savepoint=False)
@@ -851,7 +845,7 @@ def add_subscriptions_backend(
     result["already_subscribed"] = dict(result["already_subscribed"])
 
     if send_new_subscription_messages:
-        if len(result["subscribed"]) <= MAX_BULK_NEW_SUBSCRIPTION_MESSAGES:
+        if len(result["subscribed"]) <= settings.MAX_BULK_NEW_SUBSCRIPTION_MESSAGES:
             send_messages_for_new_subscribers(
                 user_profile=user_profile,
                 subscribers=subscribers,
diff --git a/zproject/default_settings.py b/zproject/default_settings.py
index 61bf1b3ad7..490a419cf4 100644
--- a/zproject/default_settings.py
+++ b/zproject/default_settings.py
@@ -681,6 +681,10 @@ MAX_STREAM_SIZE_FOR_TYPING_NOTIFICATIONS = 100
 # be soft-reactivated in the case of user group mention.
 MAX_GROUP_SIZE_FOR_MENTION_REACTIVATION = 11
 
+# The maximum number of newly subscribed users for which the server
+# will consider sending DMs to each new subscriber.
+MAX_BULK_NEW_SUBSCRIPTION_MESSAGES = 100
+
 # Limiting guest access to other users via the
 # can_access_all_users_group setting makes presence queries much more
 # expensive. This can be a significant performance problem for
```

timabbott commented on web/src/add_subscribers_pill.ts:
```diff
@@ -134,13 +143,15 @@ export function create({
     with_add_button: boolean;
     onPillCreateAction?: (pill_user_ids: number[]) => void;
     onPillRemoveAction?: (pill_user_ids: number[]) => void;
+    add_btn_pill_update_callback?: () => void;
```
Don't use abbreviations like `btn`; we use `button` as consistently as we can. (There's some legacy code).

``` diff
diff --git a/web/src/add_subscribers_pill.ts b/web/src/add_subscribers_pill.ts
index 9ab76a4c3d..67860ff0a7 100644
--- a/web/src/add_subscribers_pill.ts
+++ b/web/src/add_subscribers_pill.ts
@@ -143,7 +143,7 @@ export function create({
     with_add_button,
     onPillCreateAction,
     onPillRemoveAction,
-    add_btn_pill_update_callback,
+    add_button_pill_update_callback,
 }: {
     $pill_container: JQuery;
     get_potential_subscribers: () => User[];
@@ -151,7 +151,7 @@ export function create({
     with_add_button: boolean;
     onPillCreateAction?: (pill_user_ids: number[]) => void;
     onPillRemoveAction?: (pill_user_ids: number[]) => void;
-    add_btn_pill_update_callback?: () => void;
+    add_button_pill_update_callback?: () => void;
 }): CombinedPillContainer {
     const pill_widget = input_pill.create<CombinedPill>({
         $container: $pill_container,
@@ -207,7 +207,7 @@ export function create({
         set_up_handlers_for_add_button_state(
             pill_widget,
             $pill_container,
-            add_btn_pill_update_callback,
+            add_button_pill_update_callback,
         );
     }
 
diff --git a/web/src/stream_edit_subscribers.ts b/web/src/stream_edit_subscribers.ts
index c76bb3f01e..ca1235c5f0 100644
--- a/web/src/stream_edit_subscribers.ts
+++ b/web/src/stream_edit_subscribers.ts
@@ -144,7 +144,7 @@ export function enable_subscriber_management({
         get_potential_subscribers,
         onPillCreateAction: update_notification_choice,
         onPillRemoveAction: update_notification_choice,
-        add_btn_pill_update_callback: update_notification_choice,
+        add_button_pill_update_callback: update_notification_choice,
         get_user_groups: user_groups.get_all_realm_user_groups,
         with_add_button: true,
     });
```

kuv2707 commented:
@timabbott I've manually tested the feature (by setting `MAX_BULK_NEW_SUBSCRIPTION_MESSAGES` to 2)and the links in the api docs. I didn't find any typo.

timabbott commented:
Merged, thanks @kuv2707!
